pkg/speaker/layer2/ndp_announcer.go

```diff
@@ -0,0 +1,280 @@
+			delete(n.ip2mac, ip)
+		}
+	}
+
+	delete(n.ipranges, name)
+}
+
+func (n *ndpAnnouncer) Size() int {
+	return len(n.ipranges)
+}
+
+func (n *ndpAnnouncer) processRequest() dropReason {
+	msg, _, nsSenderIP, err := n.conn.ReadFrom()
+	if err != nil {
+		if err == io.EOF {
+			klog.Infof("ndp speaker interface %s closed", n.intf.Name)
+			return dropReasonClosed
+		}
+		klog.Errorf("ndp speaker read interface %s error: %v", n.intf.Name, err)
+		return dropReasonError
+	}
+
+	ns, ok := msg.(*ndp.NeighborSolicitation)
+	if !ok {
+		return dropReasonNotNeighborSolicitation
+	}
+
+	var nsSenderHwAddr net.HardwareAddr
+	for _, o := range ns.Options {
+		nsHwAddr, ok := o.(*ndp.LinkLayerAddress)
+		if !ok || nsHwAddr.Direction != ndp.Source {
+			continue
+		}
+		nsSenderHwAddr = nsHwAddr.Addr
+		break
+	}
+	if nsSenderHwAddr == nil {
+		return dropReasonNotSourceDirection
+	}
+
+	hwAddr := n.getMac(ns.TargetAddress.String())
+	if hwAddr == nil {
+		return dropReasonUnknowTargetIP
+	}
+
+	metrics.UpdateRequestsReceivedMetrics(ns.TargetAddress.String())
+	klog.Infof("interface %s got NDP request, send response %s-%s to %s-%s .",
+		n.intf.Name, ns.TargetAddress, *hwAddr, nsSenderIP, nsSenderHwAddr)
+
+	na := generateNDP(false, *hwAddr, ns.TargetAddress)
+	if err := n.conn.WriteTo(na, nil, nsSenderIP); err != nil {
+		klog.Errorf("failed to send neighbor advertisement packet: %v", err)
+		return dropReasonError
+	}
+	metrics.UpdateResponsesSentMetrics(ns.TargetAddress.String())
+	return dropReasonNone
+}
+
+func (n *ndpAnnouncer) getMac(ip string) *net.HardwareAddr {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	result, ok := n.ip2mac[ip]
+	if !ok {
+		return nil
+	}
+	return &result
+}
+
+func (n *ndpAnnouncer) setMac(ip string, mac net.HardwareAddr) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	if _, ok := n.ip2mac[ip]; !ok {
+		metrics.InitLayer2Metrics(ip)
+	}
+	n.ip2mac[ip] = mac
+}
+
+func generateNDP(gratuitous bool, srcHW net.HardwareAddr, srcIP netip.Addr) *ndp.NeighborAdvertisement {
+	na := &ndp.NeighborAdvertisement{
+		Router:        false,
+		Solicited:     !gratuitous,
+		Override:      gratuitous,
+		TargetAddress: srcIP,
+		Options: []ndp.Option{
+			&ndp.LinkLayerAddress{
+				Direction: ndp.Target,
+				Addr:      srcHW,
+			},
+		},
+	}
+
+	return na
+}
+
+func (n *ndpAnnouncer) Gratuitous(ip netip.Addr) error {
+	if n.getMac(ip.String()) != nil {
+		return nil
+	}
+
+	n.setMac(ip.String(), n.intf.HardwareAddr)
+	klog.Infof("store ingress ip related node ip and mac. %s-%s", ip.String(), n.intf.HardwareAddr.String())
+
+	addr, err := netip.ParseAddr(net.IPv6linklocalallnodes.String())
+	if err != nil {
+		return fmt.Errorf("parse IPv6linklocalallnodes: %v", err)
+	}
+
+	klog.Infof("send gratuitous ndp packet: %s-%s", ip, n.intf.HardwareAddr)
+	na := generateNDP(true, n.intf.HardwareAddr, ip)
+	return n.conn.WriteTo(na, nil, addr)
+}
+
+const (
+	dropReasonNotNeighborSolicitation = 6
+	dropReasonNotSourceDirection      = 7
+)
```

Comment:
You can move these, along with some other related constants, into a new const.go file.